Signal/src/views/GalleryRailView.swift

```diff
@@ -0,0 +1,274 @@
+
+    // MARK: Subview Helpers
+
+    var aspectRatioConstraint: NSLayoutConstraint?
+    func adjustAspectRatio(isSelected: Bool) {
+        if let oldConstraint = aspectRatioConstraint {
+            NSLayoutConstraint.deactivate([oldConstraint])
+        }
+
+        if isSelected, let itemAspectRatio = item?.aspectRatio {
+            aspectRatioConstraint = imageView.autoPin(toAspectRatio: itemAspectRatio)
+        } else {
+            // Portrait mode AR by default
+            let kDefaultAspectRatio: CGFloat = 9.0 / 16.0
+            aspectRatioConstraint = imageView.autoPin(toAspectRatio: kDefaultAspectRatio)
+        }
+    }
+
+    let imageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleAspectFill
+
+        return imageView
+    }()
+}
+
+protocol GalleryRailViewDelegate: class {
+    func galleryRailView(_ galleryRailView: GalleryRailView, didTapItem imageRailItem: GalleryRailItem)
+}
+
+class GalleryRailView: UIView, GalleryRailCellViewDelegate {
+
+    weak var delegate: GalleryRailViewDelegate?
+
+    // MARK: Initializers
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        addSubview(scrollView)
+        scrollView.layoutMargins = .zero
+        scrollView.autoPinEdgesToSuperviewMargins()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: Public
+
+    public func configure(itemProvider: GalleryRailItemProvider?, focusedItem: GalleryRailItem?) {
+        let animationDuration: TimeInterval = 0.2
+
+        guard let itemProvider = itemProvider else {
+            UIView.animate(withDuration: animationDuration) {
+                self.isHidden = true
+            }
+            return
+        }
+
+        let sameRailItems = { (lhs: [GalleryRailItem], rhs: [GalleryRailItem]) -> Bool in
+            guard lhs.count == rhs.count else {
+                return false
+            }
+            for (index, element) in lhs.enumerated() {
+                guard element === rhs[index] else {
+                    return false
+                }
+            }
+            return true
+        }
+
+        if itemProvider === self.itemProvider, sameRailItems(itemProvider.railItems, self.cellViewItems) {
+            UIView.animate(withDuration: animationDuration) {
+                self.updateFocusedItem(focusedItem)
+                self.layoutIfNeeded()
+            }
+        }
+
+        self.itemProvider = itemProvider
+        scrollView.subviews.forEach { $0.removeFromSuperview() }
+
+        guard itemProvider.railItems.count > 1 else {
+            UIView.animate(withDuration: animationDuration) {
+                self.isHidden = true
+            }
+            return
+        }
+
+        UIView.animate(withDuration: animationDuration) {
+            self.isHidden = false
+        }
+
+        let cellViews = buildCellViews(items: itemProvider.railItems)
```

Comment:
So we always maintain a `GalleryRailCellView` for each asset in the the album, but not for each asset in the entire "all media gallery".  

Each `GalleryRailCellView` eagerly loads a thumbnail image.  This seems like it will work fine if our max album size is 32, but it won't scale to album sizes of 1000.  I suppose we could use a UICollectionView for horizontal scrolling if we needed to?

Reply:
Right - I considered using an embedded collectionview, but I thought given that our limited album size allowed it, that it was simpler to do it this way.